num/dualcmplx/dual.go

```diff
@@ -0,0 +1,269 @@
+	return Number{
+		Real: x.Real * y.Real,
+		Dual: x.Real*y.Dual + x.Dual*cmplx.Conj(y.Real),
+	}
+}
+
+// Inv returns the dual inverse of d.
+func Inv(d Number) Number {
+	return Number{
+		Real: 1 / d.Real,
+		Dual: -d.Dual / (d.Real * cmplx.Conj(d.Real)),
+	}
+}
+
+// Conj returns the conjugate of d₁+d₂ϵ, d̅₁+d₂ϵ.
+func Conj(d Number) Number {
+	return Number{
+		Real: cmplx.Conj(d.Real),
+		Dual: d.Dual,
+	}
+}
+
+// Scale returns d scaled by f.
+func Scale(f float64, d Number) Number {
+	return Number{Real: complex(f, 0) * d.Real, Dual: complex(f, 0) * d.Dual}
+}
+
+// Abs returns the absolute value of d.
+func Abs(d Number) dual.Number {
+	return dual.Number{Real: cmplx.Abs(d.Real)}
+}
+
+// PowReal returns d**p, the base-d exponential of p.
+//
+// Special cases are (in order):
+//	PowReal(NaN+xϵ, ±0) = 1+NaNϵ for any x
+//	Pow(0+xϵ, y) = 0+Infϵ for all y < 1.
+//	Pow(0+xϵ, y) = 0 for all y > 1.
+//	PowReal(x, ±0) = 1 for any x
+//	PowReal(1+xϵ, y) = 1+xyϵ for any y
+//	Pow(Inf, y) = +Inf+NaNϵ for y > 0
+//	Pow(Inf, y) = +0+NaNϵ for y < 0
+//	PowReal(x, 1) = x for any x
+//	PowReal(NaN+xϵ, y) = NaN+NaNϵ
+//	PowReal(x, NaN) = NaN+NaNϵ
+//	PowReal(-1, ±Inf) = 1
+//	PowReal(x+0ϵ, +Inf) = +Inf+NaNϵ for |x| > 1
+//	PowReal(x+yϵ, +Inf) = +Inf for |x| > 1
+//	PowReal(x, -Inf) = +0+NaNϵ for |x| > 1
+//	PowReal(x, +Inf) = +0+NaNϵ for |x| < 1
+//	PowReal(x+0ϵ, -Inf) = +Inf+NaNϵ for |x| < 1
+//	PowReal(x, -Inf) = +Inf-Infϵ for |x| < 1
+//	PowReal(+Inf, y) = +Inf for y > 0
+//	PowReal(+Inf, y) = +0 for y < 0
+//	PowReal(-Inf, y) = Pow(-0, -y)
+func PowReal(d Number, p float64) Number {
+	switch {
+	case p == 0:
+		switch {
+		case cmplx.IsNaN(d.Real):
+			return Number{Real: 1, Dual: cmplx.NaN()}
+		case d.Real == 0, cmplx.IsInf(d.Real):
+			return Number{Real: 1}
+		}
+	case p == 1:
+		if cmplx.IsInf(d.Real) {
+			d.Dual = cmplx.NaN()
+		}
+		return d
+	case math.IsInf(p, 1):
+		if d.Real == -1 {
+			return Number{Real: 1, Dual: cmplx.NaN()}
+		}
+		if Abs(d).Real > 1 {
+			if d.Dual == 0 {
+				return Number{Real: cmplx.Inf(), Dual: cmplx.NaN()}
+			}
+			return Number{Real: cmplx.Inf(), Dual: cmplx.Inf()}
+		}
+		return Number{Real: 0, Dual: cmplx.NaN()}
+	case math.IsInf(p, -1):
+		if d.Real == -1 {
+			return Number{Real: 1, Dual: cmplx.NaN()}
+		}
+		if Abs(d).Real > 1 {
+			return Number{Real: 0, Dual: cmplx.NaN()}
+		}
+		if d.Dual == 0 {
+			return Number{Real: cmplx.Inf(), Dual: cmplx.NaN()}
+		}
+		return Number{Real: cmplx.Inf(), Dual: cmplx.Inf()}
+	case math.IsNaN(p):
+		return Number{Real: cmplx.NaN(), Dual: cmplx.NaN()}
+	case d.Real == 0:
+		if p < 1 {
+			return Number{Real: d.Real, Dual: cmplx.Inf()}
+		}
+		return Number{Real: d.Real}
+	case cmplx.IsInf(d.Real):
+		if p < 0 {
+			return Number{Real: 0, Dual: cmplx.NaN()}
+		}
+		return Number{Real: cmplx.Inf(), Dual: cmplx.NaN()}
+	}
+	return Pow(d, Number{Real: complex(p, 0)})
+}
+
+// Pow returns d**p, the base-d exponential of p.
+func Pow(d, p Number) Number {
```

Comment:
I don't see why we can do this in the dual numbers.... it might be true, but it is not obvious to me.

Reply:
This is essentially Defn 5.3.

Reply:
Huh, yeah you're right. Good enough for me, I'll figure out how exactly that works later at some point. Makes sense in principle though, so yeah. I suppose it does work. That's pretty neat.